test(docs-infra): add unit tests for rxjs examples

sonukapoor opened the pull request:
This commit adds missing unit tests for all rxjs examples from the docs.

Closes #28017

## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [X] Documentation content changes



## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No

gkalpak commented on aio/content/examples/rx-library/src/error-handling.ts:
```diff
@@ -1,25 +1,40 @@
-  }),
-  catchError(err => of([]))
-);
+export function docRegionApiData(console, ajax) {
```
I would prefer to keep the convention that these function are named after the PascalCased docregion name, prefixed with `docRegion`. For the unnamed, default docregions, I suggest we use a consistent name acriss all files. Something like `docRegionDefault()` or `docRegionUnnamed()`.

gkalpak commented on aio/content/examples/rx-library/src/operators.2.ts:
```diff
@@ -1,16 +1,26 @@
-import { Observable, of } from 'rxjs';
+// #docplaster
 
+import { of } from 'rxjs';
```
I know this was like this before, but I think this should actually be part of the docregion.

gkalpak commented on aio/content/examples/rx-library/src/operators.ts:
```diff
@@ -1,20 +1,29 @@
-
+// #docplaster
+
+import { of } from 'rxjs';
```
I know this was like this before, but I think this should actually be part of the docregion.

gkalpak commented on aio/content/examples/rx-library/src/simple-creation.1.spec.ts:
```diff
@@ -0,0 +1,12 @@
+describe('simple-creation.1', () => {
+  it('should create a promise from an observable and return an empty object', () => {
+    const console = {log: jasmine.createSpy('log')};
+    const fetch = () => of({});
```
Let's use a property to make this more distinct:

```suggestion
    const fetch = () => of({foo: 42});
```

gkalpak commented on aio/content/examples/rx-library/src/simple-creation.1.spec.ts:
```diff
@@ -0,0 +1,12 @@
+    const console = {log: jasmine.createSpy('log')};
+    const fetch = () => of({});
+    docRegionPromise(console, fetch);
+    expect(console.log).toHaveBeenCalledWith({});
+    expect(console.log).toHaveBeenCalledWith('Completed');
```
```suggestion
    expect(console.log.calls.allArgs()).toEqual([
      [{}],
      ['Completed'],
    ]);
```

gkalpak commented on aio/content/examples/rx-library/src/simple-creation.2.spec.ts:
```diff
@@ -0,0 +1,19 @@
+import { docRegionInterval } from './simple-creation.2';
+
+describe('simple-creation.2', () => {
+  beforeEach(() => {
+    jasmine.clock().install();
+  });
+
+  afterEach(() => {
+    jasmine.clock().uninstall();
+  });
```
Let's keep this consistent with other recently added tests:

```suggestion
  beforeEach(() => jasmine.clock().install());
  afterEach(() => jasmine.clock().uninstall());
```

gkalpak commented on aio/content/examples/rx-library/src/simple-creation.2.spec.ts:
```diff
@@ -0,0 +1,19 @@
+  it('should create an Observable that will publish a value on an interval', () => {
+    const console = {log: jasmine.createSpy('log')};
+    const subscription = docRegionInterval(console);
+    jasmine.clock().tick(1001);
```
```suggestion
    jasmine.clock().tick(1000);
```

Also, to make sure it works as expected, let's add an assertion that there is nothing emitted after 999ms.

gkalpak commented on aio/content/examples/rx-library/src/simple-creation.2.spec.ts:
```diff
@@ -0,0 +1,19 @@
+    const console = {log: jasmine.createSpy('log')};
+    const subscription = docRegionInterval(console);
+    jasmine.clock().tick(1001);
+    expect(console.log).toHaveBeenCalledWith('It\'s been 0 seconds since subscribing!');
```
Let's verify at least one more emission:

```suggestion
    expect(console.log).toHaveBeenCalledWith('It\'s been 0 seconds since subscribing!');
    
    console.log.calls.reset();
    
    jasmine.clock().tick(999);
    expect(console.log).not.toHaveBeenCalled();
    
    jasmine.clock().tick(1);
    expect(console.log).toHaveBeenCalledWith('It\'s been 1 seconds since subscribing!');
```

gkalpak commented on aio/content/examples/rx-library/src/simple-creation.2.ts:
```diff
@@ -0,0 +1,22 @@
+  const secondsCounter = interval(1000);
+  // Subscribe to begin publishing values
+  const subscription = secondsCounter.subscribe(n =>
+    console.log(`It's been ${n} seconds since subscribing!`));
```
This is actually wrong :smiley:

```suggestion
    console.log(`It's been ${n + } seconds since subscribing!`));
```

sonukapoor replied:
I am not following on this one. Can you please explain?

gkalpak replied:
Sorry, I meant `n + 1`. I've updated the original comment.

gkalpak commented on aio/content/examples/rx-library/src/simple-creation.ts:
```diff
@@ -61,5 +8,3 @@ const apiData = ajax('/api/data');
 apiData.subscribe(res => console.log(res.status, res.response));
```
No tests for this? 🤨

gkalpak commented on aio/content/examples/rx-library/src/simple-creation.3.spec.ts:
```diff
@@ -0,0 +1,42 @@
+import { docRegionEvent } from './simple-creation.3';
+
+describe('simple-creation.3', () => {
+  it('should call unsubscribe when clientX and clientY are below < 40 ', () => {
```
Please add two tests: One that verifies it prints coorsidnates and one that verifies the unsubscription.

sonukapoor replied:
There are many expectations that verify the coordinates through the `console.log` spy. Is there any specific we need? I will add the un-subscription test.

gkalpak replied:
We need to have small focused tests. We can test everything in one giant tests, but that is not very ergonomic when the test starts failing :smiley:

gkalpak commented on aio/content/examples/rx-library/src/simple-creation.3.spec.ts:
```diff
@@ -0,0 +1,42 @@
+    // Ensure that we have unsubscribed.
+    triggerMousemove({ clientX: 30, clientY: 30 });
+    triggerMousemove({ clientX: 50, clientY: 50 });
+    expect(console.log).not.toHaveBeenCalled();
```
Let's also verify that `removeEventListener` was called (with appropriate arguments).

gkalpak commented on aio/content/examples/rx-library/src/retry-on-error.ts:
```diff
@@ -1,26 +1,36 @@
-import { Observable, of } from 'rxjs';
-
-
+import { of } from 'rxjs';
```
I know this was like this before, but I think this should actually be part of the docregion.

gkalpak commented on aio/content/examples/rx-library/src/retry-on-error.ts:
```diff
@@ -1,26 +1,36 @@
+      }
+      return res.response;
+    }),
+    retry(3), // Retry up to 3 times before failing
```
This fix is independent from adding tests. Please move it to a separate commit.

gkalpak commented on aio/content/examples/rx-library/src/retry-on-error.spec.ts:
```diff
@@ -0,0 +1,17 @@
+      ['Error occured. Retrying...'],
+      ['data: ', []]
+    ]);
+  });
```
We should also verify that `ajax()` is called three times (perhaps by adding a `console.log()` call inside the fake ajax implementation.

sonukapoor replied:
Please correct me if I am wrong, but I don't think the `ajax()` method will be called three times. I think the underlying `retry` will `subscribe` to the returned observable from the `ajax` call and call `next` on it. I have tried to add `console.log()` into the `ajax()` method, and it gets only called once. I can add that if you like.

gkalpak replied:
Yes, you are right. Sorry for the confusion.
What I meant is verify that the observable returned by `ajax()` will be subscribed to 3 times.

You can do this with something like:

```ts
const ajax = jasmine.createSpy('ajax').and.callFake((url: string) =>
  of(true).pipe(tap(() => console.log('Subscribed to AJAX'))));
```

gkalpak commented on aio/content/examples/rx-library/src/retry-on-error.spec.ts:
```diff
@@ -0,0 +1,17 @@
+import { docRegionApiData } from './retry-on-error';
+
+describe('retry-on-error', () => {
+  it('should return an empty array', () => {
```
It should also be cool to have a test that shows that the outer observable completes with the correct value if the response comes successfully after a couple of attempts.

sonukapoor replied:
I am not entirely sure how this would be done here. Can you help?

gkalpak replied:
Sure. You could do something like:

```ts
let i = 0;
const ajax = jasmine.createSpy('ajax').and.callFake((url: string) =>
  of(null).pipe(mergeMap(() => (++i < 4) ? throwError('Test error') : of(/* Proper response */))));
```

sonukapoor replied:
I was not able to get this working, so I left it as is. Please add a commit if you can. Thanks 👍

gkalpak replied:
For reference: I pushed commit e0e139daeb4a21b19cf57fcfd597b59c7556fccd.

gkalpak reviewed (changes requested):
Thx, @sonukappor!

Awesome to see that these tests provide immediate value by finding broken examples (such as the `retry-on-error` one) :tada:

Please update the commit message to mention that this PR is targeting the `rx-library` example specifically.

sonukapoor commented on aio/content/examples/rx-library/src/error-handling.spec.ts:
```diff
@@ -0,0 +1,14 @@
+import { docRegionApiData } from './error-handling';
+
+describe('error-handling', () => {
+  it('should return an empty array', () => {
```
That makes a lot more sense.

sonukapoor commented on aio/content/examples/rx-library/src/error-handling.ts:
```diff
@@ -1,25 +1,40 @@
+      return res.response;
+    }),
+    catchError(err => {
+      console.log('Caught error.');
```
I have reverted this.

gkalpak commented on aio/content/examples/rx-library/src/simple-creation.2.ts:
```diff
@@ -0,0 +1,22 @@
+  const secondsCounter = interval(1000);
+  // Subscribe to begin publishing values
+  const subscription = secondsCounter.subscribe(n =>
+    console.log(`It's been ${n} seconds since subscribing!`));
```
As per https://github.com/angular/angular/pull/38905#discussion_r491486940, this should actually be:

```suggestion
    console.log(`It's been ${n + 1} seconds since subscribing!`));
```

gkalpak commented on aio/content/examples/rx-library/src/error-handling.ts:
```diff
@@ -1,25 +1,36 @@
-  next(x) { console.log('data: ', x); },
-  error(err) { console.log('errors already caught... will not run'); }
-});
+export function docRegionApiData(console, ajax) {
```
As per https://github.com/angular/angular/pull/38905#discussion_r491483373, I would prefer to keep the convention that these function are named after the PascalCased docregion name, prefixed with `docRegion`. For the unnamed, default docregions, I suggest we use a consistent name acriss all files. Something like `docRegionDefault()` or `docRegionUnnamed()`.

gkalpak commented on aio/content/examples/rx-library/src/simple-creation.3.spec.ts:
```diff
@@ -0,0 +1,42 @@
+
+    // Ensure that we have unsubscribed.
+    triggerMousemove({ clientX: 30, clientY: 30 });
+    expect(input.removeEventListener).toHaveBeenCalledWith('mousemove', jasmine.any(Function), undefined);
```
Couldn't this be:

```suggestion
    expect(input.removeEventListener).toHaveBeenCalledWith('mousemove', triggerMousemove, undefined);
```

gkalpak commented on aio/content/examples/rx-library/src/simple-creation.3.spec.ts:
```diff
@@ -0,0 +1,42 @@
+import { docRegionEvent } from './simple-creation.3';
+
+describe('simple-creation.3', () => {
+  it('should call unsubscribe when clientX and clientY are below < 40 ', () => {
```
As per https://github.com/angular/angular/pull/38905#discussion_r491490796, this should be broken up into multiple tests (i.e. `it()` calls).

gkalpak commented on aio/content/examples/rx-library/src/simple-creation.ts:
```diff
@@ -1,65 +1,20 @@
+  const apiData = ajax('/api/data');
+  // Subscribe to create the request
+  apiData.subscribe(res => console.log(res.status, res.response));
+
```
There is not reason to have an empty line here (since it is at the end of the docregion). Whitespace is trimmed on docregion content anyway.

gkalpak commented on aio/content/examples/rx-library/src/retry-on-error.ts:
```diff
@@ -1,20 +1,28 @@
+/* tslint:disable:align */
 // #docregion
-
+import { of } from 'rxjs';
 import { ajax } from 'rxjs/ajax';
 import { map, retry, catchError } from 'rxjs/operators';
```
Please indent the imports (as stated by the comment above and as is done in other examples) and indent the function body (for consistency).

gkalpak commented on aio/content/examples/rx-library/src/retry-on-error.ts:
```diff
@@ -1,20 +1,28 @@
-  map(res => {
+  map((res: any) => {
     if (!res.response) {
+      console.log('Error occured. Retrying...');
```
The `Retrying...` part is not necessarily true and therefore confusing.
Let's stick to:

```suggestion
      console.log('Error occured.');
```

gkalpak reviewed (changes requested):
Reminder to self: https://github.com/angular/angular/pull/38905#discussion_r491494196 is still pending.

gkalpak reviewed (changes requested):
https://github.com/angular/angular/pull/38905#discussion_r491963927 doesn't seem to have been addressed.

Reminder to self: https://github.com/angular/angular/pull/38905#discussion_r491494196 is still pending.

sonukapoor commented on aio/content/examples/rx-library/src/error-handling.spec.ts:
```diff
@@ -0,0 +1,47 @@
+
+    docRegionDefault(mockConsole, ajax);
+
+    subject.next({});
```
Good catch!

gkalpak commented on aio/content/examples/rx-library/src/retry-on-error.spec.ts:
```diff
@@ -0,0 +1,49 @@
+
+  it('should return an empty array after 3 retries + 1 initial request', () => {
+    const ajax = jasmine.createSpy('ajax').and.callFake((url: string) => {
+      return of(true).pipe(tap(() => mockConsole.log('Subscribed to AJAX')));
```
`true` is a little confusing. Let's use `of({})` instead.

gkalpak reviewed (changes requested):
https://github.com/angular/angular/pull/38905#discussion_r494813072 doesn't seem to have been addressed.

##
Reminder to self: https://github.com/angular/angular/pull/38905#discussion_r491494196 is still pending.

sonukapoor commented:
> [#38905 (comment)](https://github.com/angular/angular/pull/38905#discussion_r494813072) doesn't seem to have been addressed.
> 
> Reminder to self: [#38905 (comment)](https://github.com/angular/angular/pull/38905#discussion_r491494196) is still pending.

I see the change here:

https://github.com/angular/angular/blob/546e665760e6acbcb1225d57c8808eb3ae319fcf/aio/content/examples/rx-library/src/retry-on-error.ts

sonukapoor commented on aio/content/examples/rx-library/src/error-handling.spec.ts:
```diff
@@ -25,21 +25,20 @@ describe('error-handling', () => {
-  it('should return an empty array when the ajax method throws an error', () => {
-    ajax = jasmine.createSpy('ajax').and.callFake((url: string) => throwError('Test Error'));
+  it('should return an empty array when the ajax observable errors', () => {
+    ajax = ajax.and.returnValue(throwError('Test Error'));
```
Good point

sonukapoor commented on aio/content/examples/rx-library/src/retry-on-error.spec.ts:
```diff
@@ -0,0 +1,49 @@
+  beforeEach(() => mockConsole = { log: jasmine.createSpy('log') });
+
+  it('should return the response object', () => {
+    const ajax = jasmine.createSpy('ajax').and.callFake((url: string) => {
+      return of({ response: { foo: 'bar' } });
+    });
```
Since we do not have any linter any place for this. What is the rule of thumb here to break the lines?

gkalpak replied:
The limit is 100 chars per line.

sonukapoor replied:
Is that documented anywhere?

gkalpak replied:
It's in [CONTRIBUTING.md#rules][1].

[1]: https://github.com/angular/angular/blob/323be39297450fa5a4a70f800511722b4964dec2/CONTRIBUTING.md#rules

gkalpak commented:
>> [#38905 (comment)](https://github.com/angular/angular/pull/38905#discussion_r494813072) doesn't seem to have been addressed.
>> 
>> Reminder to self: [#38905 (comment)](https://github.com/angular/angular/pull/38905#discussion_r491494196) is still pending.

> I see the change here:
>
> https://github.com/angular/angular/blob/546e665760e6acbcb1225d57c8808eb3ae319fcf/aio/content/examples/rx-library/src/retry-on-error.ts

I don't see it. My comment was about `error-handling.spec.ts`, but you linked to `retry-on-error.ts` 😕

sonukapoor commented:
> I don't see it. My comment was about `error-handling.spec.ts`, but you linked to `retry-on-error.ts` 😕

Strange, pretty sure that the original comment was pointing to the import alignment. Regardless, I have fixed the variable renaming from `subject` to `ajaxSubject`.

sonukapoor committed:
fix(docs-infra): fix the retry-on-error example

Previously, the `retry` example did not work as intended. The `retry`
operator was called before the exception occured, thus not retrying the
`ajax` request.

This commit moves the `retry` operator into the correct order to ensure
that the failed request is retried.

sonukapoor committed:
test(docs-infra): add unit tests for rx-library examples

This commit adds missing unit tests for all rx-library examples from the docs.

Closes #28017

gkalpak committed:
fixup! test(docs-infra): add unit tests for rx-library examples

gkalpak committed:
test(docs-infra): add missing test for the `rx-library` docs example

This commit adds an extra test for the `retry-on-error` snippet of the
`rx-library` docs example to ensure it can successfully recover after a
couple of failed attempts.

This commit addresses comment
https://github.com/angular/angular/pull/38905#discussion_r491494196.

gkalpak commented:
@googlebot I consent.

gkalpak reviewed (approved):
LGTM :shipit:
Thank you, @sonukapoor, for getting everything sorted out ❤️ 

Reviewed-for: global-docs-approvers

gkalpak commented:
@sonukapoor: FYI, I rebased this on master, squashed fixup commits, added a new fixup commit with some minor clean-up and added a new commit with the test we discussed in https://github.com/angular/angular/pull/38905#discussion_r491494196.

Please, take a look and confirm this looks good to you 👀

sonukapoor commented:
> @sonukapoor: FYI, I rebased this on master, squashed fixup commits, added a new fixup commit with some minor clean-up and added a new commit with the test we discussed in [#38905 (comment)](https://github.com/angular/angular/pull/38905#discussion_r491494196).
> 
> Please, take a look and confirm this looks good to you 👀

LGTM. Thanks @gkalpak
